src/normal.cc

```diff
@@ -88,11 +88,17 @@ template<InsertMode mode>
 void enter_insert_mode(Context& context, NormalParams)
 {
     context.input_handler().insert(mode);
+
```

Comment:
I am not really keen on any solution that requires us to add code/wrap every single function that might end up trying to modify the buffer. I'd prefer a solution which will, by design, work. I think throwing a runtime error on tentative to modify a read only buffer could make sense.